File: src/main/java/com/dude/dms/backend/brain/pdf/PdfToDocParser.java
package com.dude.dms.backend.brain.pdf;

import com.dude.dms.backend.brain.BrainUtils;
import com.dude.dms.backend.data.base.Doc;
import com.dude.dms.backend.data.base.Tag;
import com.dude.dms.backend.data.rules.PlainTextRule;
import com.dude.dms.backend.data.rules.RegexRule;
import com.dude.dms.backend.service.DocService;
import com.dude.dms.backend.service.PlainTextRuleService;
import com.dude.dms.backend.service.RegexRuleService;
import com.dude.dms.backend.service.TagService;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;
import java.util.stream.Collectors;

import static com.dude.dms.backend.brain.OptionKey.*;

@Component
public class PdfToDocParser implements Parser {

    private static final Logger LOGGER = LoggerFactory.getLogger(PdfToDocParser.class);

    @Autowired
    private DocService docService;

    @Autowired
    private TagService tagService;

    @Autowired
    private PlainTextRuleService plainTextRuleService;

    @Autowired
    private RegexRuleService regexRuleService;

    private final String docSavePath;

    public PdfToDocParser() {
        docSavePath = BrainUtils.getProperty(DOC_SAVE_PATH);
    }

    /**
     * Creates a new {@link Doc} from a file.
     *
     * @param file the file used for parsing, this has a to be a pdf file.
     */
    public void parse(File file) {
        LOGGER.info("Parsing file {}...", file.getName());
        String guid = UUID.randomUUID().toString();
        File savedFile = saveFile(file, guid + ".pdf");

        Doc doc = new Doc(guid);

        String rawText = stripText(savedFile);
        if (rawText != null && !rawText.isEmpty()) {
            doc.setRawText(rawText);
            doc.setDocumentDate(discoverDates(rawText));
        }

        doc.setTags(discoverTags(rawText));

        docService.create(doc);
        LOGGER.info("Created doc with ID: {}", doc.getGuid());
    }

    /**
     * Tries to find tags for a doc by all available methods
     * @param rawText raw text of the document
     * @return list of tags
     */
    private Set<Tag> discoverTags(String rawText) {
        Set<Tag> tags = discoverAutoTags();
        tags.addAll(discoverRuleTags(rawText));
        return tags;
    }

    /**
     * Validates all active rules against the raw text and adds tags for matching rules.
     * @param rawText raw text
     * @return list of tags
     */
    private Set<Tag> discoverRuleTags(String rawText) {
        Set<Tag> tags = new HashSet<>();

        for (PlainTextRule rule : plainTextRuleService.getActiveRules()) {
            for (String line : rawText.split("\n")) {
                if (rule.validate(line)) {
                    Set<Tag> ruleTags = tagService.findByPlainTextRule(rule);
                    LOGGER.info("{} found a match! Adding tags[{}]...", rule, ruleTags.stream().map(Tag::getName).collect(Collectors.joining(",")));
                    tags.addAll(ruleTags);
                    break;
                }
            }
        }

        for (RegexRule rule : regexRuleService.getActiveRules()) {
            for (String line : rawText.split("\n")) {
                if (rule.validate(line)) {
                    Set<Tag> ruleTags = tagService.findByRegexRule(rule);
                    LOGGER.info("{} found a match! Adding tags[{}]...", rule, ruleTags.stream().map(Tag::getName).collect(Collectors.joining(",")));
                    tags.addAll(ruleTags);
                    break;
                }
            }
        }
        return tags;
    }

    /**
     * Gets tags, wich are set to automatic in the user_settings
     * @return list of tags
     */
    private Set<Tag> discoverAutoTags() {
        Set<Tag> tags = new HashSet<>();
        if (Boolean.parseBoolean(BrainUtils.getProperty(AUTO_REVIEW_TAG))) {
            tagService.findById(Long.parseLong(BrainUtils.getProperty(REVIEW_TAG_ID))).ifPresent(tag -> {
                LOGGER.info("Adding tag: {}", tag.getName());
                tags.add(tag);
            });
        }
        return tags;
    }

    private static String stripText(File file) {
        String rawText = null;
        try (PDDocument pdDoc = PDDocument.load(file)) {
            PDFTextStripper pdfStripper = new PDFTextStripper();
            LOGGER.info("Stripping text...");
            rawText = pdfStripper.getText(pdDoc);
        } catch (IOException e) {
            LOGGER.error(e.getMessage());
        }
        return rawText;
    }

    private File saveFile(File file, String newName) {
        Path targetPath = Paths.get(docSavePath, newName);
        LOGGER.info("Copying to {}...", targetPath);
        try {
            Files.move(file.toPath(), targetPath, StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException e) {
            LOGGER.error(e.getMessage());
        }
        return new File(targetPath.toUri());
    }

    private static LocalDate discoverDates(String rawText) {
        LOGGER.info("Trying to find date...");
        String[] datePatterns = BrainUtils.getProperty(DATE_SCAN_FORMATS).split(",");
        Map<LocalDate, Integer> map = new HashMap<>();
        for (String pattern : datePatterns) {
            for (String line : rawText.split("\n")) {
                for (int i = 0; i < (line.length() - pattern.length()); i++) {
                    String snippet = line.substring(i, i + pattern.length());
                    try {
                        LocalDate date = LocalDate.parse(snippet, DateTimeFormatter.ofPattern(pattern));
                        map.put(date, map.getOrDefault(date, 0) + 1);
                    } catch (DateTimeParseException ignored) {
                    }
                }
            }
        }
        Optional<Map.Entry<LocalDate, Integer>> entry = map.entrySet().stream().max(Map.Entry.comparingByValue());
        if (entry.isPresent()) {
            LOGGER.info("Setting date as {} with {} occurences", entry.get().getKey(), entry.get().getValue());
            return entry.get().getKey();
        }
        LOGGER.info("No date found on document");
        return null;
    }
}

File: src/main/java/com/dude/dms/brain/parsing/PdfToDocParser.kt
package com.dude.dms.brain.parsing

import com.dude.dms.brain.DmsLogger
import com.dude.dms.brain.FileManager
import com.dude.dms.brain.ParseEvent
import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.data.docs.Doc
import com.dude.dms.backend.data.docs.TextBlock
import com.dude.dms.backend.service.DocService
import com.dude.dms.backend.service.TagService
import com.dude.dms.backend.service.TextBlockService
import com.dude.dms.brain.options.Options
import org.apache.pdfbox.pdmodel.PDDocument
import org.springframework.stereotype.Component
import java.io.File
import java.io.IOException
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.*

@Component
class PdfToDocParser(
        private val docService: DocService,
        private val tagService: TagService,
        private val textBlockService: TextBlockService,
        private val plainTextRuleValidator: PlainTextRuleValidator,
        private val regexRuleValidator: RegexRuleValidator,
        private val fileManager: FileManager,
        private val pdfStripper: DmsPdfTextStripper
) : Parser {

    private val textBlockList = mutableListOf<TextBlock>()

    private val eventListeners = HashMap<String, ParseEvent>()

    /**
     * Creates a new [Doc] from a file.
     *
     * @param file the file used for parsing, this has a to be a pdf file.
     */
    override fun parse(file: File) {
        LOGGER.info("Parsing file {}...", file.name)
        val guid = file.name.takeWhile { it != '.' }
        var rawText: String? = null
        try {
            PDDocument.load(file).use { pdDoc ->
                fileManager.saveImage(pdDoc, guid)
                rawText = stripText(pdDoc)
            }
            val doc = Doc(guid)
            if (!rawText.isNullOrEmpty()) {
                doc.rawText = rawText
                doc.documentDate = discoverDates(rawText!!)
            }
            doc.tags = discoverTags(rawText).toMutableSet()
            docService.create(doc)
            LOGGER.info("Created doc with ID: {}", doc.guid)
            textBlockList.forEach {
                it.doc = doc
                textBlockService.create(it)
                LOGGER.info("Created textblock {} for doc {}", it, doc)
            }
            eventListeners.values.forEach { it.invoke(true) }
        } catch (e: IOException) {
            e.message?.let { LOGGER.error(it) }
            eventListeners.values.forEach { it.invoke(false) }
        }
    }

    /**
     * Tries to find tags for a doc by all available methods
     *
     * @param rawText raw text of the document
     * @return list of tags
     */
    private fun discoverTags(rawText: String?): Set<Tag> {
        val tags: MutableSet<Tag> = HashSet()
        for (tag in Options.get().tag.automaticTags) {
            tagService.findByName(tag)?.let {
                LOGGER.info("Adding tag: {}", it.name)
                tags.add(it)
            }
        }
        if (!rawText.isNullOrEmpty()) {
            tags.addAll(plainTextRuleValidator.getTags(rawText))
            tags.addAll(regexRuleValidator.getTags(rawText))
        }
        return tags
    }

    private fun stripText(pdDoc: PDDocument): String {
        LOGGER.info("Stripping text...")
        textBlockList.clear()
        return pdfStripper.getTextWithPositions(pdDoc, textBlockList as ArrayList<TextBlock>)
    }

    fun addEventListener(key: String, parseEvent: ParseEvent) {
        eventListeners[key] = parseEvent
    }

    private fun discoverDates(rawText: String): LocalDate? {
        LOGGER.info("Trying to find date...")
        val map = mutableMapOf<LocalDate, Int>()
        for (pattern in Options.get().view.dateScanFormats) {
            for (line in rawText.split("\n").toTypedArray()) {
                for (i in 0 until line.length - pattern.length) {
                    val snippet = line.substring(i, i + pattern.length)
                    try {
                        val date = LocalDate.parse(snippet, DateTimeFormatter.ofPattern(pattern))
                        map[date] = map.getOrDefault(date, 0) + 1
                    } catch (ignored: DateTimeParseException) {
                    }
                }
            }
        }
        map.entries.maxBy { it.value }?.let {
            LOGGER.info("Setting date as {} with {} occurences", it.key, it.value)
            return it.key
        }
        LOGGER.info("No date found on document")
        return null
    }

    companion object {
        private val LOGGER = DmsLogger.getLogger(PdfToDocParser::class.java)
    }
}